### PR TITLE
Enable all (most) of the warnings as errors

### PR DESCRIPTION
**Pull request by oerdnj**
Enable -Wall -Wextra and couple of additional warnings and treat
them as errors.

Because the implementation is not complete, disable unused parameter
warning.

And fix all issues found.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -54,6 +54,26 @@ AC_TYPE_SIZE_T
...
+
+AX_CHECK_COMPILE_FLAG([-fno-strict-aliasing],
+		      [STD_CFLAGS="$STD_CFLAGS -fno-strict-aliasing"])
+AX_CHECK_COMPILE_FLAG([-Werror -fno-delete-null-pointer-checks],
+		      [STD_CFLAGS="$STD_CFLAGS -fno-delete-null-pointer-checks"])
+AX_CHECK_COMPILE_FLAG([-fdiagnostics-show-option],
+		      [STD_CFLAGS="$STD_CFLAGS -fdiagnostics-show-option"])
+
+AC_SUBST([STD_CFLAGS])
+
```
Comment (simo5) [inline configure.ac:76]: This should stay near the check for c11, either all up to it, or move the c11 cheack down here...

File: src/provider.h [hunk 2]
```diff
@@ -18,7 +18,7 @@
 #include <openssl/proverr.h>
 #include <openssl/core_names.h>
 
-#define UNUSED __attribute__((unused))
+#define UNUSED __attribute__((__unused__))
```
Comment (simo5) [inline src/provider.h:21]: Why this change ?

File: src/asymmetric_cipher.c [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 #include "provider.h"
 #include <string.h>
+#include <strings.h>
```
Comment (simo5) [inline src/asymmetric_cipher.c:6]: this came from the strcasecmp) usage right?
If so drop from here and the other two files

File: src/kdf.c [hunk 4]
```diff
@@ -298,7 +298,7 @@ static int p11prov_hkdf_set_ctx_params(void *ctx, const OSSL_PARAM params[])
     /* can be multiple paramaters, which wil be all concatenated */
     for (p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_INFO); p != NULL;
          p = OSSL_PARAM_locate_const(p + 1, OSSL_KDF_PARAM_INFO)) {
-        void *ptr;
+        uint8_t *ptr;
```
Comment (simo5) [inline src/kdf.c:301]: Why this change ?

File: src/provider.h (hunk 2 above)
Comment (oerdnj) [reply src/provider.h:21]: There was issue in the past with the standard `noreturn` (from `#include <stdnoreturn.h>`) which mangled system headers that use `__attribute__((noreturn))` and mangled it to `__attribute__((_Noreturn))` which in turn `clang` ignored, so it's kind of habit to use the underscored variants of attributes, especially in headers.

There's no other reason here, so this could be dropped, I did this on autopilot.

File: src/kdf.c (hunk 4 above)
Comment (simo5) [reply src/kdf.c:301]: Might be better to cast later if it is a type issue.

Comment (simo5) [reply src/kdf.c:301]: Thinking again, it is probably ok, nvm

File: src/keymgmt.c [hunk 5]
```diff
@@ -165,7 +165,8 @@ static int p11prov_rsakm_secbits(int bits)
     if (bits > 4096) return 152;
     if (bits > 3072) return 128;
     if (bits > 2048) return 112;
-    if (bits <= 2048) return 0;
```
Comment (simo5) [inline src/keymgmt.c:168]: I think I will have to fix these comparisons, they probably should all be >=, returning 0 for 2048 is wrong ...

File: src/kdf.c (hunk 4 above)
Comment (oerdnj) [reply src/kdf.c:301]: Pointer arithmetics on `void *` is GNU extension and not defined by C standard.

File: configure.ac (hunk 1 above)
Comment (oerdnj) [reply configure.ac]: Done.

File: src/provider.h (hunk 2 above)
Comment (oerdnj) [reply src/provider.h]: Removed.

File: src/kdf.c (hunk 4 above)
Comment (simo5) [reply src/kdf.c:301]: Yeah, I realized that when I expanded the code below, I already forgotten this code :-D

File: src/provider.h (hunk 2 above)
Comment (simo5) [reply src/provider.h]: As lon as I know the reason I am fin with it

File: src/provider.c [hunk 6]
```diff
@@ -312,9 +312,14 @@ static int p11prov_get_capabilities(void *provctx, const char *capability,
...
 
+struct reason_string {
+    unsigned int id;
+    const char *ptr;
+};
+
 static const OSSL_ITEM *p11prov_get_reason_strings(void *provctx)
 {
-    static const OSSL_ITEM reason_strings[] = {
+    static const struct reason_string reason_strings[] = {
```
Comment (simo5) [inline src/provider.c:322]: Why this change?

Comment (oerdnj) [reply src/provider.c:338]: The `OSSL_ITEM` is `struct { int, void * }` and the compiler complains about dropping the `const` modifier when assigning `const char *` to `void *`.

It was either this or typecasting all strings in the array to `char *` (which also would be a valid fix). I am fine with either, but this is less chatty.

File: src/keymgmt.c (hunk 5 above)
Comment (oerdnj) [reply src/keymgmt.c]: I did this while rebasing and fixing the conflicts, but if you want this as a separate commit, that's fine too.

File: src/asymmetric_cipher.c (hunk 3 above)
Comment (oerdnj) [reply src/asymmetric_cipher.c]: Yeah, it had the extra commits from other branches which are now merged, so I dropped this.

File: src/provider.c (hunk 6 above)
Comment (simo5) [reply src/provider.c:338]: yeah lets do a
#define C(str) (void *)(str)
before the struct and call them all as C("string")

I rather not shadow structures defined in headers.

Comment (oerdnj) [reply src/provider.c]: Done

Comment (simo5) [reply src/provider.c]: perfect
